SIGSEGV with socket_writer on shutdown if could not write

## Bug report

### Relevant telegraf.conf:
```toml
[[outputs.socket_writer]]
  address = "udp://127.0.0.1:8099"
```

### System info:

6e76731b

### Steps to reproduce:

1. Start Telegraf with nothing listening to udp address
2. Send SIGINT to initiate shutdown

### Expected behavior:

Clean shutdown

### Actual behavior:
```
2017-05-02T01:17:15Z E! Error writing to output [socket_writer]: write udp 127.0.0.1:56307->127.0.0.1:8099: write: connection refused
^C2017-05-02T01:17:15Z I! Hang on, flushing any cached metrics before shutdown
2017-05-02T01:17:15Z E! Error writing to output [socket_writer]: write udp 127.0.0.1:34613->127.0.0.1:8099: write: connection refused
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1052c5c]

goroutine 1 [running]:
github.com/influxdata/telegraf/plugins/outputs/socket_writer.(*SocketWriter).Close(0xc4200198c0, 0x1aff960, 0xc42014c100)
        <autogenerated>:3 +0x3c
github.com/influxdata/telegraf/agent.(*Agent).Close(0xc42016c040, 0x13f19e0, 0xc42014ff80)
        /home/dbn/src/influx/src/github.com/influxdata/telegraf/agent/agent.go:77 +0x66
github.com/influxdata/telegraf/agent.(*Agent).Run(0xc42016c040, 0xc420168cc0, 0x0, 0x0)
        /home/dbn/src/influx/src/github.com/influxdata/telegraf/agent/agent.go:401 +0x4c4
main.reloadLoop(0xc4201687e0, 0x1c0b6b0, 0x0, 0x0, 0x1c0b6b0, 0x0, 0x0, 0x1c0b6b0, 0x0, 0x0, ...)
        /home/dbn/src/influx/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:222 +0xa06
main.main()
        /home/dbn/src/influx/src/github.com/influxdata/telegraf/cmd/telegraf/telegraf.go:392 +0x65b
```
### Additional info:

--- a/plugins/outputs/socket_writer/socket_writer.go
+++ b/plugins/outputs/socket_writer/socket_writer.go
@@ -124,6 +124,16 @@ func (sw *SocketWriter) Write(metrics []telegraf.Metric) error {
 	return nil
 }
 
+// Close closes the connection. Noop if already closed.
+func (sw *SocketWriter) Close() error {
+	if sw.Conn == nil {
+		return nil
+	}
+	err := sw.Conn.Close()
+	sw.Conn = nil
+	return err
+}
+
 func newSocketWriter() *SocketWriter {
 	s, _ := serializers.NewInfluxSerializer()
 	return &SocketWriter{

--- a/plugins/outputs/socket_writer/socket_writer_test.go
+++ b/plugins/outputs/socket_writer/socket_writer_test.go
@@ -143,7 +143,7 @@ func TestSocketWriter_Write_err(t *testing.T) {
 
 	// close the socket to generate an error
 	lconn.Close()
-	sw.Close()
+	sw.Conn.Close()
 	err = sw.Write(metrics)
 	require.Error(t, err)
 	assert.Nil(t, sw.Conn)

